go/ql/lib/semmle/go/security/UnicodeBypassValidationQuery.qll

```diff
@@ -0,0 +1,85 @@
+ * Provides a taint-tracking configuration for detecting "Unicode transformation mishandling" vulnerabilities.
+ */
+
+private import go
+import semmle.go.Concepts
+import semmle.go.dataflow.DataFlow
+import semmle.go.dataflow.TaintTracking
+import UnicodeBypassValidationCustomizations::UnicodeBypassValidation
+
+/** A state signifying that a logical validation has not been performed. */
+class PreValidation extends DataFlow::FlowState {
+  PreValidation() { this = "PreValidation" }
+}
+
+/** A state signifying that a logical validation has been performed. */
+class PostValidation extends DataFlow::FlowState {
+  PostValidation() { this = "PostValidation" }
+}
+
+/**
+ * A taint-tracking configuration for detecting "Unicode transformation mishandling" vulnerabilities.
+ *
+ * This configuration uses two flow states, `PreValidation` and `PostValidation`,
+ * to track the requirement that a logical validation has been performed before the Unicode Transformation.
+ */
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "UnicodeBypassValidation" }
+
+  override predicate isSource(DataFlow::Node source, DataFlow::FlowState state) {
+    (source instanceof UntrustedFlowSource or source instanceof Source) and
+    state instanceof PreValidation
+  }
+
+  override predicate isAdditionalTaintStep(
+    DataFlow::Node nodeFrom, DataFlow::FlowState stateFrom, DataFlow::Node nodeTo,
+    DataFlow::FlowState stateTo
+  ) {
+    (
+      exists(DataFlow::CallNode cn |
+        cn.getACalleeIncludingExternals().asFunction() instanceof EscapeFunction and
+        nodeFrom = cn.getAnArgument() and
+        nodeTo = cn.getResult()
+      )
+      or
+      exists(DataFlow::CallNode cn |
+        (
+          cn.getACalleeIncludingExternals().asFunction() instanceof RegexpMatchFunction or
+          cn.getACalleeIncludingExternals().asFunction() instanceof RegexpReplaceFunction
+        ) and
+        nodeFrom = cn.getAnArgument() and
+        nodeTo = cn.getResult()
+      )
+      or
+      exists(DataFlow::CallNode cn |
+        cn.getTarget()
+            .hasQualifiedName("strings",
+              [
+                "Clone", "Compare", "Contains", "ContainsAny", "ContainsRune", "Count", "Cut",
+                "CutPrefix", "CutSuffix", "EqualFold", "Fields", "FieldsFunc", "HasPrefix",
+                "HasSuffix", "Index", "IndexAny", "IndexByte", "IndexFunc", "IndexRune", "Join",
+                "LastIndex", "LastIndexAny", "LastIndexByte", "LastIndexFunc", "Map", "Repeat",
+                "Replace", "ReplaceAll", "Split", "SplitAfter", "SplitAfterN", "SplitN", "Title",
+                "ToLower", "ToLowerSpecial", "ToTitle", "ToTitleSpecial", "ToUpper",
+                "ToUpperSpecial", "ToValidUTF8", "Trim", "TrimFunc", "TrimLeft", "TrimLeftFunc",
+                "TrimPrefix", "TrimRight", "TrimRightFunc", "TrimSpace", "TrimSuffix"
+              ]) and
+        nodeFrom = cn.getAnArgument() and
+        nodeTo = cn
```

Comment:
Use getResult etc -- and note that the Cut functions return multiple results, so using `cn` as the result won't work.

Reply:
How about this commit [8cc03c6](https://github.com/github/codeql/pull/12994/commits/8cc03c6f85626d9a22e25bccc9a405617feacab9) ? I'm getting better right ?

Reply:
Yes, the usage of getResult is better for sure! For the ones that don't produce a result, such as `Index`, I think you probably need a different kind of query -- you'd be looking for an `Index` check that guards unicode normalisation, since in this case the data flow doesn't travel through the check, but rather the check and the unicode norming are related by control flow and use of a common variable. Take a look at `RedirectCheckBarrier` for example to illustrate using `DataFlow::BarrierGuard` to characterise situations like

```go
if strings.Index(x, "<") { norm.NFC.String(x) }
```

I suggest adding a test case that uses these sorts of checks.

Reply:
`BarrierGuard` is a bit difficult concept to grasp. 

If I understood well, the example code is not a full barrier guard.

```go
if strings.Index(x, "<") == -1 { norm.NFC.String(x) }
``` 

But something like may be:

```go
if !strings.ContainsAny(x, "<﹤＜") { norm.NFKC.String(x) }
``` 

This is what I did so far. 
1. I have added a test which somehow still bypassable

```go
unicode_input := req.URL.Query().Get("unicode_input")
		escaped := html.EscapeString(unicode_input)
		if strings.Index(escaped, "<") == -1 {
			unicode_norm := norm.NFKC.String(escaped) // $result=BAD
			fmt.Println(w, "Results: %q", unicode_norm)
		}
```

2. I removed `Index`-like calls from being additional taint steps to become barrier guard.

```ql
private predicate untrustedUnicodeCharCheckGuard(DataFlow::Node cn, Expr e, boolean outcome) {
  cn.(DataFlow::CallNode)
      .getTarget()
      .hasQualifiedName("strings",
        [
          "Contains", "ContainsAny", "ContainsRune", "Count", "EqualFold", "HasPrefix", "HasSuffix",
          "Index", "IndexAny", "IndexByte", "IndexFunc", "IndexRune", "LastIndex", "LastIndexAny",
          "LastIndexByte", "LastIndexFunc",
        ]) and
  cn.(DataFlow::CallNode).getArgument(0).asExpr() = e and
  outcome = true
}

/**
 * A call to a function called `Index`, `ContainsAny`, or similar, which may be
 * considered a barrier guard for eliminating untrusted Unicode characters.
 */
class UntrustedUnicodeCharCheckBarrier extends DataFlow::Node {
  UntrustedUnicodeCharCheckBarrier() {
    this = DataFlow::BarrierGuard<untrustedUnicodeCharCheckGuard/3>::getABarrierNode()
  }
}
```

3. Added a sanitizer

```ql
 override predicate isSanitizer(DataFlow::Node sanitizer, DataFlow::FlowState state) {
    (
      sanitizer instanceof UntrustedUnicodeCharCheckBarrier or
      sanitizer instanceof Sanitizer
    ) and
    state instanceof PreValidation
  }
```

What am I missing please? what do you think needs to be updated for sure?

Reply:
Hi @Sim4n6! I answered a question about barrier guards [in response to someone else's question](https://github.com/github/codeql/discussions/13343#discussioncomment-6089089) just yesterday which might help you as well. 

It seems like your understand of barrier guards based on the two examples you showed is correct. I think one problem with your implementation and the reason that your test doesn't work as expected is that not all of the functions you are checking for in `untrustedUnicodeCharCheckGuard` actually return booleans. For functions like `Index`, you would need to do a bit more work to check that the result is what you'd expect for validation to pass. As an example for just `Index` which checks that the result is compared to `-1`:

```ql
private predicate indexCheck(DataFlow::Node g, Expr e, boolean outcome) {
  exists(DataFlow::CallNode cn, DataFlow::EqualityTestNode etn |
    g = etn and
    DataFlow::localFlow(cn.getResult(), etn.getAnOperand()) and
    cn.getTarget().hasQualifiedName("strings", "Index") and
    cn.getArgument(0).asExpr() = e and
    etn.getAnOperand().getIntValue() = -1 and
    outcome = etn.getPolarity()
  )
}
```

(You can adapt this for all methods where the result should be compared to `-1`.)

We can now add this into `untrustedUnicodeCharCheckGuard` for testing: 

```ql
private predicate untrustedUnicodeCharCheckGuard(DataFlow::Node cn, Expr e, boolean outcome) {
  indexCheck(cn, e, outcome) or
  (
    cn.(DataFlow::CallNode)
    .getTarget()
    .hasQualifiedName("strings",
      [
        "Contains", "ContainsAny", "ContainsRune", "Count", "EqualFold", "HasPrefix", "HasSuffix",
        "IndexAny", "IndexByte", "IndexFunc", "IndexRune", "LastIndex", "LastIndexAny",
        "LastIndexByte", "LastIndexFunc",
      ]) and
    cn.(DataFlow::CallNode).getArgument(0).asExpr() = e and
    outcome = true
  )
}
```

If I perform a quick evaluation of `isSanitizer` with these changes, then I get a result for the `bad2` test case, as I think is expected. Though, at this point the overall query still returns a result for the `bad2` test case. I think that is because you check for `state instanceof PreValidation` in `isSanitizer`. Changing that to `state instanceof PostValidation` gives the expected results for the current tests (i.e. removes the result for `bad2`) since you want the sanitizer to apply post escaping but before use.

Reply:
> I answered a question about barrier guards [in response to someone else's question](https://github.com/github/codeql/discussions/13343#discussioncomment-6089089) just yesterday which might help you as well.

Lucky me :)


> It seems like your understand of barrier guards based on the two examples you showed is correct. I think one problem with your implementation and the reason that your test doesn't work as expected is that not all of the functions you are checking for in `untrustedUnicodeCharCheckGuard` actually return booleans. 

I have updated to `intCheck` and `boolCheck` predicates. I just tweaked it a little bit.

> If I perform a quick evaluation of `isSanitizer` with these changes, then I get a result for the `bad2` test case, as I think is expected. Though, at this point the overall query still returns a result for the `bad2` test case. Changing that to `state instanceof PostValidation` gives the expected results for the current tests (i.e. removes the result for `bad2`) since you want the sanitizer to apply post escaping but before use.

My problem was that the Golang snippet code checks were a partial implementation and the sample `bad2` is a good one for handling Unicode variant of `<`.  

>  I think that is because you check for `state instanceof PreValidation` in `isSanitizer`. 

I missed that one. 

Now the resulting query supports BarrierGuard. That would reduces the false positives but also misses some true ones. Say we have a code like this one `if strings.Index(x, "<") == -1 { norm.NFC.String(x) }`, according to the query this is a valid barrier, but this is bypassable with a variant of `"<"`. Voila.

Reply:
I'd suggest varying this somewhat by using the guard to condition your sinks, rather than using it as a sanitizer.

In principle, this would look something like (pseudocode):

```go
x := getUntrusted()
y := strings.Replace(x, ">" "_") // y gets state PostValidation
sink(y)

...

x2 := getUntrusted()
if (!strings.Contains(x2, ">")) {
  sink(x2) // x2 is still in state PreValidation, but is covered by a relevant guard
}
```

So this could look something like:

```ql
predicate isSink(DataFlow::Node n, DataFlow::FlowState state) {
  n = getANormArgument() and
  (
    state instanceof PostValidation // n has had sensitive characters replaced already
    or
    n instanceof UntrustedUnicodeCharCheckBarrier // n is covered by a condition like `if !n.Contains("<")`, so should be considered a sink whether or not we've seen `Replace`-like calls.
}
```

Reply:
I would certainly recommend adding tests cases covering these variations on the query by the way, so we have clear concrete examples to work on.

Reply:
sure

Reply:
Done, I have added this _bad_ sample code:

```go	http.HandleFunc("/bad2", func(w http.ResponseWriter, req *http.Request) {

		unicode_input := req.URL.Query().Get("unicode_input")
		if !strings.Contains(unicode_input, "<") {
			unicode_norm := norm.NFKC.String(unicode_input) // $Missing:result=BAD
			fmt.Println(w, "Results: %q", unicode_norm)
		}

	})
````

This code won't be detected due to the barrier guards. So I have added `$Missing:result=BAD` test tag.

Reply:
ah, nice trick @smowton , you mean UntrustedUnicodeCharCheckBarriers are somehow a sinking matter. 
And the samples are all bad, let me update that right away.